OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,282 @@
+    virtual ~AbstractDataTable()                             = default;
+
+    /** Get metadata associated with the table.                               */
+    const TableMetaData& getTableMetaData() const {
+        return _tableMetaData;
+    }
+
+    /** Update metadata associated with the table.                            */
+    TableMetaData& updTableMetaData() {
+        return _tableMetaData;
+    }
+
+    /** Get number of rows.                                                   */
+    size_t getNumRows() const {
+        return implementGetNumRows();
+    }
+
+    /** Get number of dependent columns.                                      */
+    size_t getNumColumns() const {
+        return implementGetNumColumns();
+    }
+    
+    /** Get metadata associated with the independent column.                  */
+    const IndependentMetaData& getIndependentMetaData() const {
+        return _independentMetaData;
+    }
+    
+    /** Set metadata associated with the independent column.                  */
+    void 
+    setIndependentMetaData(const IndependentMetaData& independentMetaData) {
+        _independentMetaData = independentMetaData;
+        validateIndependentMetaData();
+    }
+
+    /** Get metadata associated with the dependent columns.                   */
+    const DependentsMetaData& getDependentsMetaData() const {
+        return _dependentsMetaData;
+    }
+    
+    /** Set metadata associated with the dependent columns.                   */
+    void 
+    setDependentsMetaData(const DependentsMetaData& dependentsMetaData) {
+        _dependentsMetaData = dependentsMetaData;
+        validateDependentsMetaData();
+    }
+
+protected:
+    /** Get number of rows. Implemented by derived classes.                   */
+    virtual size_t implementGetNumRows() const       = 0;
+    
+    /** Get number of columns. Implemented by derived classes.                */
+    virtual size_t implementGetNumColumns() const    = 0;
+    
+    /** Check if metadata for independent column is valid . Implemented by 
+    derived classes.                                                          */
+    virtual void validateIndependentMetaData() const = 0;
+
+    /** Check if metadata for dependent column is valid. Implemented by derived
+    classes.                                                                  */
+    virtual void validateDependentsMetaData() const  = 0;
+
+    TableMetaData       _tableMetaData;
+    DependentsMetaData  _dependentsMetaData;
+    IndependentMetaData _independentMetaData;
+}; // AbstractDataTable
+
+
+/** DataTable_ is a in-memory storage container for data with support for 
+holding metadata (using the base class AbstractDataTable). Data contains an 
+independent column and a set of dependent columns. The type of the independent 
+column can be configured using ETX (template param). The type of the dependent 
+columns, which together form a matrix, can be configured using ETY (template 
```

Comment:
I have added some description of dependent and independent columns to AbstractDataTable.